### PR TITLE
LIVY-342. Create Livy UI: Create Web UI Servlet and All Sessions Page

### PR DESCRIPTION
**Pull request by ajbozarth**
This is the first PR creating a Livy UI for monitoring current Livy Sessions and Batches.

In this PR I have laid down the groundwork for the Livy UI. I created a new UIServlet and added it to the Livy server. I added a resource servlet for hosting html, css, png and js files for the UI. Finally using html, css, and js files I created a summary page containing tables listing all current interactive sessions and batches retrieved from their respective GET calls to the REST API.

Note: the Livy logo is a shrunken version of the logo in the gh-pages branch.

This passes current unit tests and ITs, I did not add new test for the UI since it's all html returns, but can try if it is requested.

Screenshots:
both
only-sessions
only-batches
empty

### CODE REVIEW

**Comment by ajbozarth** [comment]
A few points of discussion:
- Should we mount the ui at /* instead of /ui/* (and if we do where should the metrics currently there go?)
- Should I add pagination? If so should I do it in a follow up JIRA?
- Are there any new style checks or tests I should add/look into/address

**Comment by ajbozarth** [comment]
@zjffdu this is from your JIRA, also @alex-the-man and @jerryshao you both have commented on this before, what do you think? This is just the first part of the UI I'm building.

**Comment by jerryshao** [comment]
I think it is great as a first step of UI things:

> Should we mount the ui at /* instead of /ui/* (and if we do where should the metrics currently there go?)

I'm incline to choose "/*", metrics could be merged as a new tab of this UI if it is easy to work.

> Should I add pagination? If so should I do it in a follow up JIRA?

In the long term I think it is necessary. At least we should leave room for the improvement later.

>Are there any new style checks or tests I should add/look into/address

What specifically are you referring to?

**Comment by ajbozarth** [comment]
I'll look into moving it to /* and adding a metrics tab (and tabs in general I guess, the nav bar can already it).

And I'll open up a new JIRA for pagination tomorrow to do later.

As for the third point, I'm pretty sure there's not any style checking happening on the html, css and js files, so should I add it. Also should I write tests for UIServlet even though all it's functions return html?

**Comment by jerryshao** [comment]
I think we could add some UI selenium tests if possible, like `UISeleniumSuite` in Spark.

File: server/src/main/scala/com/cloudera/livy/LivyConf.scala [hunk 1]
```diff
@@ -64,6 +64,8 @@ object LivyConf {
   val SERVER_PORT = Entry("livy.server.port", 8998)
   val CSRF_PROTECTION = LivyConf.Entry("livy.server.csrf-protection.enabled", false)
 
+  val UI_ENABLED = Entry("livy.ui.enabled", false)
```
Comment (zjffdu) [inline server/src/main/scala/com/cloudera/livy/LivyConf.scala]: Add this into livy.conf

**Comment by zjffdu** [comment]
Thanks @ajbozarth It's awesome UI.  Several comments :
* Link the application Id to yarn app if it is yarn mode
* Pagination for the sessions
* Retention for the session, maybe only display the last 100 sessions.

**Comment by ajbozarth** [comment]
@jerryshao I'll add a followup JIRA to create a SeleniumSuite, it's been a while since I wrote Selenium tests so I'm looking forward to that.

@zjffdu I'll look at adding that yarn link, and the SessionId will link to a session detail page in the next PR so I'll look at adding tooltips to clarify the links appropriately as well. And I'll get pagination done here rather than a followup JIRA

@zjffdu For the latest point I actually have a question, I'm filling the tables by calling GET /sessions or /batches and I'm not sure what the retention is on those returns. If it isn't retaining apps properly we should address it there, from what I can tell it does remove apps from that list but I'm not sure when and under what conditions since I still see dead sessions in my tables during testing.

**Comment by ajbozarth** [comment]
General PR Question for @alex-the-man or @zjffdu : Is there a way to get Travis CI to rerun and similarly to get Codecov to run? They don't seem to behave consistently and I'm not sure what calls them.

**Comment by ajbozarth** [comment]
tl;dr next steps from comments:
- Move from /ui/* to /*
- Add tabbing to navbar
- Add Metrics tab to replace current /* content
- Paginate tables
- Link AppId to YARN App
- Add tooltip clarify AppId link
- Add style checking (probably open a new JIRA)
- Create Selenium tests (open a new JIRA)

**Comment by ajbozarth** [comment]
So quick update on some issues I've run into address the above list:

- Moving from `/ui/*` to `/*` might be an issue. In spark the UI is essentially at `/*` and the REST API at some kind of `/api/*` in the case of livy the server is currently just the REST API and so it's located at `/*` so if we decide to also put the UI at `/*` we would have to avoid any paths used in the API, such as `/sessions/*` and `/batches/*` which could end up making the UI paths non-intuitive (for instance my plan was to put a session summary at `/ui/sessions/:id`). Since Livy is a REST API first I think the API should stay at `/*` and we should put the UI at `/ui/*` as I already have. I can however have `/` redirect to `/ui/` and create a ui tab that displays the json data returned in the metrics servlet (which I'll be moving to `/metrics`)
- As I mentioned I'll be moving metrics from `/` to `/metrics` and grabbing data from it for a UI metrics tab, that's because the servlet is from a library and can't be imbedded into the UIServlet..

So unless theres any arguments, I plan to leave the UI at `/ui/*` and use redirects for usability.

**Comment by zjffdu** [comment]
@ajbozarth You can close and reopen this PR to trigger the CI. Or you can make a dummy commit to trigger the rerun.

```
git commit --amend
git push origin -f
```

**Comment by ajbozarth** [comment]
As mentioned before I've left the ui at `/ui/*` and added a redirect from `/` moving the metrics dash to `/metrics`. After some fiddling I decided adding the metrics into the UI was clunky and left it on it's own. 

Also during testing I realized the No Sessions message was broken and I fixed that.

I'm looking into pagination, but I believe it will take jQuery DataTables and may be better in a followup PR like I will be doing with js style checking and Selenium tests.

Lastly I haven't been able to figure out how to get the yarn app url to add the appId link, but I am still looking into it.

**Comment by ajbozarth** [comment]
After a bit more looking adding a link to the yarn app is non-trivial and would be better placed in a followup JIRA.

Similarly in an effort to move this forward I will be opening a followup JIRA for pagination as well.

**Comment by ajbozarth** [comment]
@alex-the-man @zjffdu @jerryshao I've addressed your feedback either in code or in a pr comment if you could take another look and see if this is good to go, if so I'll move on to the next PR.

**Comment by ajbozarth** [comment]
Opened followup JIRAS:
https://issues.cloudera.org/browse/LIVY-352
https://issues.cloudera.org/browse/LIVY-353
https://issues.cloudera.org/browse/LIVY-354

**Comment by jerryshao** [comment]
@ajbozarth can you please post a latest screenshot of your UI?

File: server/src/main/scala/com/cloudera/livy/server/LivyServer.scala [hunk 2]
```diff
@@ -142,6 +144,20 @@ class LivyServer extends Logging {
       }
     }
 
+    // Servlet for hosting static files such as html, css, and js
+    // Necessary since Jetty cannot set it's resource base inside a jar
+    val staticResourceServlet = new ScalatraServlet {
+      get("/*") {
```
Comment (jerryshao) [inline server/src/main/scala/com/cloudera/livy/server/LivyServer.scala]: should it be "/ui/"?

File: server/src/main/scala/com/cloudera/livy/server/LivyServer.scala [hunk 3]
```diff
@@ -167,7 +183,14 @@ class LivyServer extends Logging {
             val batchServlet = new BatchSessionServlet(batchSessionManager, sessionStore, livyConf)
             mount(context, batchServlet, "/batches/*")
 
-            context.mountMetricsAdminServlet("/")
+            if (livyConf.getBoolean(UI_ENABLED)) {
+              val uiServlet = new UIServlet(livyConf)
+              mount(context, uiServlet, "/ui/*")
+              mount(context, staticResourceServlet, "/static/*")
+              mount(context, uiRedirectServlet, "/*")
```
Comment (jerryshao) [inline server/src/main/scala/com/cloudera/livy/server/LivyServer.scala]: Will this also redirect REST APIs? I saw you use "/*".

**Review by jerryshao — COMMENTED** [review]
Overall LGTM, why not enable UI by default?

File: server/src/main/scala/com/cloudera/livy/server/LivyServer.scala (hunk 2 above)
Comment (ajbozarth) [inline server/src/main/scala/com/cloudera/livy/server/LivyServer.scala]: I wanted to separate the static file hosting servlet from the ui servlet, this was just a preference on my part. I felt that `UIServlet` calls should only return html for consistency, whereas the static resources are files

File: server/src/main/scala/com/cloudera/livy/server/LivyServer.scala (hunk 3 above)
Comment (ajbozarth) [inline server/src/main/scala/com/cloudera/livy/server/LivyServer.scala]: It does not, I tried just using `"/"` but the `mount` function experts a wildcard in that String

**Comment by ajbozarth** [comment]
I left the UI configurable and off by default in case it gets large by the time I'm done. Since the Livy Server itself is meant to be lightweight I didn't think having it spin up a UI every time would always be the best idea. If, as someone whose used Livy longer, you think this isn't an issue I'll remove the config and if statement.

**Comment by ajbozarth** [comment]
Also I can take new screenshots tomorrow but the only visual change that's occurred is the `h4` text is shifted 5px to the right (`padding-left=5px`)

**Comment by jerryshao** [comment]
I would like to enable UI by default, I think it is quite useful for monitoring, especially for some users who heavily rely on Livy to proxy Spark applications. Also the UI is just a bunch of JS, we don't open a new port for UI, I don't think it will bring in lots of overhead.

**Comment by jerryshao** [comment]
@zjffdu what do you think?

**Comment by zjffdu** [comment]
Agree with @jerryshao I don't see any strong reason to disable it.

**Comment by ajbozarth** [comment]
I'm ok with enabling by default, do we still want it to be configurable so it can be disabled or should I just remove the config altogether?

**Comment by ajbozarth** [comment]
@zjffdu @jerryshao After some work I have a commit ready with pagination, I can either push that here or if you think this is good to go then we can merge this as is and I will open up a new PR with that code. I'll wait to push until I hear from you.

**Comment by ajbozarth** [comment]
Also @alex-the-man I haven't heard any feedback from you, you sounded interested on slack.

File: server/src/main/resources/com/cloudera/livy/server/ui/static/all-sessions.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+function tdWrap(str) {
+    return "<td>" + str + "</td>";
```
Comment (jerryshao) [inline server/src/main/resources/com/cloudera/livy/server/ui/static/all-sessions.js]: nit: 2 space indent.

File: server/src/main/scala/com/cloudera/livy/server/ui/UIServlet.scala [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+package com.cloudera.livy.server.ui
+
+import scala.xml.Node
+
+import org.scalatra.ScalatraServlet
+
+import com.cloudera.livy.LivyConf
+
+class UIServlet(livyConf: LivyConf) extends ScalatraServlet {
```
Comment (jerryshao) [inline server/src/main/scala/com/cloudera/livy/server/ui/UIServlet.scala]: livyConf is not used here.

File: server/src/main/scala/com/cloudera/livy/server/LivyServer.scala [hunk 6]
```diff
@@ -167,7 +183,12 @@ class LivyServer extends Logging {
             val batchServlet = new BatchSessionServlet(batchSessionManager, sessionStore, livyConf)
             mount(context, batchServlet, "/batches/*")
 
-            context.mountMetricsAdminServlet("/")
+            val uiServlet = new UIServlet(livyConf)
```
Comment (jerryshao) [inline server/src/main/scala/com/cloudera/livy/server/LivyServer.scala]: I think it would be better to make UI configurable and enable it by default.

**Review by jerryshao — COMMENTED** [review]
LGTM, I think it is good to go as the first patch. We could achieve other things in separate PRs, that would be easy to review compared to one big PR.

**Comment by jerryshao** [comment]
Besides I think you need to pass the travis check first.

File: server/src/main/resources/com/cloudera/livy/server/ui/static/all-sessions.js (hunk 4 above)
Comment (ajbozarth) [inline server/src/main/resources/com/cloudera/livy/server/ui/static/all-sessions.js]: nice catch, my auto-style had a 4 space indent when I original wrote the js/html/css files, must've missed this when I updated it.

File: server/src/main/scala/com/cloudera/livy/server/ui/UIServlet.scala (hunk 5 above)
Comment (ajbozarth) [inline server/src/main/scala/com/cloudera/livy/server/ui/UIServlet.scala]: I assumed it would be needed eventually, but I can remove it for now and add it in later when it's actually needed.

File: server/src/main/scala/com/cloudera/livy/server/LivyServer.scala (hunk 6 above)
Comment (ajbozarth) [inline server/src/main/scala/com/cloudera/livy/server/LivyServer.scala]: I'll put the config back in then

**Comment by ajbozarth** [comment]
@jerryshao I addressed your comments as well as a couple bugs I found while working on the pagination that I'll be adding in the next pr. Hopefully CI will pass this time, I'm not sure why the ITs failed before.

**Comment by ajbozarth** [comment]
I tracked down why the ITs were failing: when starting up the Livy Server in ITs it checks for a 200 response from `/` but with my change from having metrics at `/` to a redirect to `/ui/` the response became a 302. To solve this I change the test to check the new metrics location. At the same time I also added an else case when the UI is disabled to redirect `/` to `/metrics` so hitting the livy server URL always returns a successful response whether the UI is enabled or not.

**Comment by jerryshao** [comment]
LGTM, thanks for the contribution.

**Comment by ajbozarth** [comment]
@zjffdu or @alex-the-man this look good to merge? I have a followup PRs ready and waiting

**Comment by zjffdu** [comment]
LGTM, will merge it

**Comment by jerryshao** [comment]
@ajbozarth , can you please fix this, this issue seems coming from UI according to my verification.

```
17/05/10 11:01:37 WARN util.MimeTypes$: There was an error detecting the mime type.
eu.medsea.mimeutil.MimeException: InputStream must support the mark() and reset() methods.
	at eu.medsea.mimeutil.MimeUtil2.getMimeTypes(MimeUtil2.java:495)
	at org.scalatra.util.Mimes$$anonfun$inputStreamMime$1.apply(Mimes.scala:54)
	at org.scalatra.util.Mimes$$anonfun$inputStreamMime$1.apply(Mimes.scala:54)
	at scala.util.control.Exception$Catch.apply(Exception.scala:102)
	at org.scalatra.util.Mimes$class.detectMime(Mimes.scala:87)
	at org.scalatra.util.Mimes$class.inputStreamMime(Mimes.scala:53)
	at org.scalatra.util.MimeTypes$.inputStreamMime(Mimes.scala:102)
	at org.scalatra.util.Mimes$class.apply(Mimes.scala:96)
	at org.scalatra.util.MimeTypes$.apply(Mimes.scala:102)
	at org.scalatra.ScalatraBase$$anonfun$contentTypeInferrer$1.applyOrElse(ScalatraBase.scala:381)
	at scala.PartialFunction$Lifted.apply(PartialFunction.scala:218)
	at scala.PartialFunction$Lifted.apply(PartialFunction.scala:214)
	at org.scalatra.ScalatraBase$class.renderResponse(ScalatraBase.scala:363)
	at org.scalatra.ScalatraServlet.renderResponse(ScalatraServlet.scala:49)
	at org.scalatra.ScalatraBase$class.executeRoutes(ScalatraBase.scala:189)
	at org.scalatra.ScalatraServlet.executeRoutes(ScalatraServlet.scala:49)
	at org.scalatra.ScalatraBase$$anonfun$handle$1.apply$mcV$sp(ScalatraBase.scala:113)
	at org.scalatra.ScalatraBase$$anonfun$handle$1.apply(ScalatraBase.scala:113)
	at org.scalatra.ScalatraBase$$anonfun$handle$1.apply(ScalatraBase.scala:113)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:57)
	at org.scalatra.DynamicScope$class.withResponse(DynamicScope.scala:80)
	at org.scalatra.ScalatraServlet.withResponse(ScalatraServlet.scala:49)
	at org.scalatra.DynamicScope$$anonfun$withRequestResponse$1.apply(DynamicScope.scala:60)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:57)
	at org.scalatra.DynamicScope$class.withRequest(DynamicScope.scala:71)
	at org.scalatra.ScalatraServlet.withRequest(ScalatraServlet.scala:49)
	at org.scalatra.DynamicScope$class.withRequestResponse(DynamicScope.scala:59)
	at org.scalatra.ScalatraServlet.withRequestResponse(ScalatraServlet.scala:49)
	at org.scalatra.ScalatraBase$class.handle(ScalatraBase.scala:111)
	at org.scalatra.ScalatraServlet.org$scalatra$servlet$ServletBase$$super$handle(ScalatraServlet.scala:49)
	at org.scalatra.servlet.ServletBase$class.handle(ServletBase.scala:43)
	at org.scalatra.ScalatraServlet.handle(ScalatraServlet.scala:49)
	at org.scalatra.ScalatraServlet.service(ScalatraServlet.scala:54)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:110)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
```

I created a Livy JIRA to track this, can you please fix it (https://issues.cloudera.org/browse/LIVY-357).

**Comment by jerryshao** [comment]
@ajbozarth have you got a chance to look at this issue I mentioned above?

**Comment by ajbozarth** [comment]
I commented on the JIRA asking for repo steps, once I know how to get it I'll find a fix

**Comment by jerryshao** [comment]
Sorry @ajbozarth , seems JIRA is broken, I cannot receive notification from JIRA.

The step is quite simple:

1. build livy with "mvn package -DskipTests"
2. Configure Spark home to use spark 2.1.0
3. Create a interactive session and open a Livy UI.

Then you will see the issue.

**Comment by ajbozarth** [comment]
Thanks, I'll look at it tomorrow. And odd about JIRA, I know the auto-tag doesn't work but I though writing out a tag manually would still send notifications. In the future would Slack be a good place to ping you if I comment on a JIRA?

**Comment by jerryshao** [comment]
Yes, you can ping me on Slack.

**Comment by prabhu1984** [comment]
Hi @ajbozarth,

Thanks for the nice UI. It's working great for non-kerberized clusters. 

But, Livy UI is not working for kerberized clusters. The web page looks empty. If we turn off auth mechanism, it works.

Any thoughts?

Thanks
Prabhu

**Comment by ajbozarth** [comment]
That's a bit surprising since that page in particular should work without authentication, but I will be adding authentication to the entire UI in a future JIRA: https://issues.cloudera.org/browse/LIVY-366

**Comment by jerryshao** [comment]
I doubt this is due to spengo authentication. If Livy Server Spnego is enabled but browser doesn't have spengo authenticated, then it will return 401 instead.

**Comment by ajbozarth** [comment]
I'll look into testing on a kerberized dev cluster and open a followup JIRA/PR once I've identified the issue and got a fix (or include it in the authentication JIRA)

**Comment by ajbozarth** [comment]
@prabhu1984 I set up and test both the code on master and my laster UI code (which I haven't pushed yet) with both a kerberized cluster and non-kerberized livy and with kerberized livy and it works perfect. I did need to do a few things to get my browser to recognize the kerberos auth though. In Safari it worked as soon as I did a kinit in my terminal, for firefox/chrome I followed some instructions I found on Google. I also noted that though I haven't addressed it directly yet, as long as the kerberos user has access control the access control won't break the website either. Based on this my best guess is that your browser was not setup correctly to work with kerberos, or was set up with the wrong user.
